source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.cc

```diff
@@ -0,0 +1,45 @@
+#include "source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.h"
+
+#include "envoy/extensions/resource_monitors/downstream_connections/v3/downstream_connections.pb.h"
+
+#include "source/common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ResourceMonitors {
+namespace DownstreamConnections {
+
+ActiveDownstreamConnectionsResourceMonitor::ActiveDownstreamConnectionsResourceMonitor(
+    const envoy::extensions::resource_monitors::downstream_connections::v3::
+        DownstreamConnectionsConfig& config)
+    : max_(config.max_active_downstream_connections()), current_(0){};
+
+bool ActiveDownstreamConnectionsResourceMonitor::tryAllocateResource(int64_t increment) {
```

Comment:
In real use cases can tryAllocateResources and tryDeallocateResources be called concurrently?

Reply:
yes, they can be called for example from multiple worker threads

Reply:
Do think using RAII for the resource allocated / decremented is probably the way to go to stop incorrect from being possible and leaking resources.

For example if max = 15, and originally we have `current=10` with some bad calls and threads swapping I think we can up with strange results:
1. Call to tryAlloc(10) that stops at line 20 (doesn't yet read current for the decrement)
2. Call to deAlloc(20) => current is still 20, so we decrement and store 0 into current
3. The first call reads current (now 0) and decrements 10, giving us current or -10 :(

Reply:
Realised that RAII/Memento approach will not work with cross thread visibility. Various threads will be accessing  tryAlloc/deAlloc resource via their thread local overload state object. The point of using atomic counter was to bypass periodic slow flushes/updates in OM (where thread local overload state is periodically updated for all threads) and instead perform faster checks from any thread relying on atomic guarantees for cross thread visibility.
@KBaichoo wdyt?